### PR TITLE
ibkr implementation

### PR DESCRIPTION
**Pull request by Ish2K**
IBKR implementation

### CODE REVIEW

File: DiscordAlertsTrader/brokerages/ibkr_api.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+                'instrument':{'symbol':order['orderLegCollection'][0]['instrument']['symbol']},
+                'instruction': order['orderLegCollection'][0]['instruction'],
+                'quantity': order['orderLegCollection'][0]['quantity'],
+            }]             
+        }
+        return order_info
+    
+    def get_quotes(self, symbol:list):
+        pass
+    def send_order(self, side:str, order_type:str, quantity:int, contract: dict, price=None, stop_price=None):
```
Comment (AdoNunes) [inline DiscordAlertsTrader/brokerages/ibkr_api.py:144]: Could you add functions that generate contracts that can be read by `send_order` with bto lims, stc lim, bracket orders and trailing stops, like these:

    def make_BTO_lim_order(self, Symbol:str, Qty:int, price:float, action="BTO", **kwarg):
        "Buy with a limit order"

        kwargs = {}
        if action == "BTO":
            kwargs['action'] = "BUY"
        elif action == "STO":
            print("STO not available for WeBull")
            return
        Symbol = self.fix_symbol(Symbol, "in")
        if "_" in Symbol:
            kwargs['asset'] = 'option'
            optionId = self.get_option_id(Symbol)
            if optionId is None:
                print("No optionId found for", Symbol)
                return None        
            kwargs['optionId'] = optionId
        else:
            kwargs['asset'] ='stock'
            kwargs['outsideRegularTradingHour'] = True
            kwargs['stock'] = Symbol

        kwargs['enforce'] ='GTC'
        kwargs['quant'] = Qty  
        kwargs['orderType'] = 'LMT' 
        kwargs['lmtPrice'] = price
        return kwargs

    def make_Lim_SL_order(self, Symbol:str, Qty:int,  PT:float, SL:None, action="STC", **kwarg):
        """Sell with a limit order and a stop loss order"""        
        kwargs = {}
        if action == "STC":
            kwargs['action'] = "SELL"
        elif action == "BTC":
            print("BTC not available for WeBull")
            return

        Symbol = self.fix_symbol(Symbol, "in")
        if "_" in Symbol:
            kwargs['asset'] = 'option'
            optionId = self.get_option_id(Symbol)
            if optionId is None:
                print("No optionId found for", Symbol)
                return None        
            kwargs['optionId'] = optionId
        else:
            kwargs['asset'] ='stock'
            kwargs['outsideRegularTradingHour'] = True
            kwargs['stock'] = Symbol

        kwargs['enforce'] ='GTC'
        kwargs['quant'] = Qty 
        if SL is not None:
            print("WARNING: webull api does not support OCO, setting a SL only, do not provide SL to send PT")
            kwargs['orderType'] = 'STP' 
            kwargs['lmtPrice'] = SL
        else:
            print("WARNING: webull api does not support OCO, sending PT without SL")
            kwargs['orderType'] = 'LMT' 
            kwargs['lmtPrice'] = PT
        return kwargs

    def make_STC_lim(self, Symbol:str, Qty:int, price:float, strike=None, action="STC", **kwarg):
        """Sell with a limit order and a stop loss order"""        
        kwargs = {}
        if action == "STC":
            kwargs['action'] = "SELL"
        elif action == "BTC":
            print("BTC not available for WeBull")
            return

        Symbol = self.fix_symbol(Symbol, "in")
        if "_" in Symbol:
            kwargs['asset'] = 'option'
            optionId = self.get_option_id(Symbol)
            if optionId is None:
                print("No optionId found for", Symbol)
                return None        
            kwargs['optionId'] = optionId
        else:
            kwargs['asset'] ='stock'
            kwargs['outsideRegularTradingHour'] = True
            kwargs['stock'] = Symbol

        kwargs['enforce'] ='GTC'
        kwargs['quant'] = Qty  
        kwargs['orderType'] = 'LMT' 
        kwargs['lmtPrice'] = price
        return kwargs

    def make_STC_SL(self, Symbol:str, Qty:int, SL:float, action="STC", **kwarg):
        """Sell with a stop loss order"""
        kwargs = {}
        if action == "STC":
            kwargs['action'] = "SELL"
        elif action == "BTC":
            print("BTC not available for WeBull")
            return

        Symbol = self.fix_symbol(Symbol, "in")
        if "_" in Symbol:
            kwargs['asset'] = 'option'
            optionId = self.get_option_id(Symbol)
            if optionId is None:
                print("No optionId found for", Symbol)
                return None        
            kwargs['optionId'] = optionId
        else:
            kwargs['asset'] ='stock'
            kwargs['outsideRegularTradingHour'] = False
            kwargs['stock'] = Symbol

        kwargs['enforce'] ='GTC'
        kwargs['quant'] = Qty 
        kwargs['orderType'] = 'STP' 
        kwargs['stpPrice'] = SL
        kwargs['lmtPrice'] = SL
        return kwargs

    def make_STC_SL_trailstop(self, Symbol:str, Qty:int,  trail_stop_const:float, action="STC", **kwarg):
        "trail_stop_const"
        kwargs = {}
        if action == "STC":
            kwargs['action'] = "SELL"
        elif action == "BTC":
            print("BTC not available for WeBull")
            return

        Symbol = self.fix_symbol(Symbol, "in")
        if "_" in Symbol:
            print("WARNING webull does not support trailing stop for options")
            return {}        

        kwargs['asset'] ='stock'
        kwargs['outsideRegularTradingHour'] = True
        kwargs['stock'] = Symbol
        kwargs['enforce'] ='GTC'
        kwargs['quant'] = Qty 
        kwargs['orderType'] = 'STP TRAIL'
        kwargs['trial_value'] = trail_stop_const
        kwargs['trial_type'] = 'DOLLAR'
        kwargs['outsideRegularTradingHour'] = True
        return kwargs

File: DiscordAlertsTrader/brokerages/ibkr_api.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+            "closeTime": closeTime,
+            'orderLegCollection':[{
+                'instrument':{'symbol':order['orderLegCollection'][0]['instrument']['symbol']},
+                'instruction': order['orderLegCollection'][0]['instruction'],
+                'quantity': order['orderLegCollection'][0]['quantity'],
+            }]             
+        }
+        return order_info
+    
+    def get_quotes(self, symbol:list):
```
Comment (AdoNunes) [inline DiscordAlertsTrader/brokerages/ibkr_api.py:142]: please add get quotes, for example request quotes for symbols = ["NFLX_051724C450", "AMD_051724C100"]

Comment (AdoNunes) [reply DiscordAlertsTrader/brokerages/ibkr_api.py:142]: maybe ibkr format convention does not match the bot formatting, like tradestation API, that would require something like that to map in and out the option formats:

    def _convert_option_fromts(self, ticker):
        """
        Convert ticker from 'SPX yearmonthdayC110' to 'SPX_monthdayyearC110' format.
        
        Parameters:
        ticker (str): Ticker in the original format.
        
        Returns:
        str: Ticker in the desired format.
        """
        if " " not in ticker:
            return ticker
        symb, date_part = ticker.split()  # Split the ticker by spaces
        formatted_date = date_part[2:6] +date_part[:2]   # Reformat the date part
        return f"{symb}_{formatted_date}{date_part[6:]}"  # Combine the parts in the desired format
    
    def _convert_option_tots(self, ticker):
            """
            Convert ticker from 'SPX_monthdayyearC110' to 'SPX yearmonthdayC110' format.
            
            Parameters:
            ticker (str): Ticker in the original format.
            
            Returns:
            str: Ticker in the desired format.
            """
            if "_" not in ticker:
                return ticker
            symb, date_part = ticker.split("_")  # Split the ticker by spaces
            formatted_date = date_part[4:6] + date_part[:4]   # Reformat the date part
            return f"{symb} {formatted_date}{date_part[6:]}"  # Combine the parts in the desired format

Comment (Ish2K) [reply DiscordAlertsTrader/brokerages/ibkr_api.py]: Done

File: DiscordAlertsTrader/brokerages/ibkr_api.py [hunk 3]
```diff
@@ -0,0 +1,525 @@
...
+        
+        Parameters:
+        ticker (str): Ticker in the original format.
+        
+        Returns:
+        dict: kwargs in desired format.
+        """
+        if "_" not in ticker:
+            return ticker
+        symb, option_part = ticker.split(" ")  # Split the ticker by spaces
```
Comment (AdoNunes) [inline DiscordAlertsTrader/brokerages/ibkr_api.py:453]: IBKR has spaces in the ticker??? are you sure you changed that example from tradestation?

File: DiscordAlertsTrader/brokerages/ibkr_api.py [hunk 4]
```diff
@@ -0,0 +1,525 @@
...
+
+        else:
+            kwargs['asset'] ='STK'
+            kwargs['outsideRegularTradingHour'] = True
+            kwargs['stock'] = Symbol
+
+        kwargs['enforce'] ='GTC'
+        kwargs['quant'] = Qty 
+        kwargs['conId'] = self.get_con_id(Symbol)
+        if SL is not None:
```
Comment (AdoNunes) [inline DiscordAlertsTrader/brokerages/ibkr_api.py:327]: IBKR allows bracket orders. That function should create a bracket order, you are not doing that

File: DiscordAlertsTrader/brokerages/ibkr_api.py [hunk 5]
```diff
@@ -0,0 +1,525 @@
...
+            kwargs['asset'] = 'OPT'
+        else:
+            kwargs['asset'] ='STK'
+            kwargs['outsideRegularTradingHour'] = False
+            kwargs['stock'] = Symbol
+
+        kwargs['enforce'] ='GTC'
+        kwargs['quant'] = Qty 
+        kwargs['orderType'] = 'STP LMT' 
+        kwargs['stpPrice'] = SL
```
Comment (AdoNunes) [inline DiscordAlertsTrader/brokerages/ibkr_api.py:381]: its a stop market order not a stop lim

File: DiscordAlertsTrader/brokerages/ibkr_api.py [hunk 6]
```diff
@@ -0,0 +1,525 @@
...
+
+        return kwargs if kwargs['conId'] is not None else None
+
+    def make_STC_SL(self, Symbol:str, Qty:int, SL:float, action="STC", **kwarg):
+        """Sell with a stop loss order"""
+        kwargs = {}
+        if action == "STC":
+            kwargs['action'] = "SELL"
+        elif action == "BTC":
+            kwargs['action'] = "BUY"
```
Comment (AdoNunes) [inline DiscordAlertsTrader/brokerages/ibkr_api.py:368]: IBRK does not have smth else like "BUYCLOSE"?

File: DiscordAlertsTrader/brokerages/ibkr_api.py (hunk 3 above)
Comment (AdoNunes) [reply DiscordAlertsTrader/brokerages/ibkr_api.py:453]: should  be _convert_option_fromIBKR

File: DiscordAlertsTrader/brokerages/ibkr_api.py [hunk 7]
```diff
@@ -0,0 +1,525 @@
...
+        strike = int(option_part[7:])
+
+        return {
+            'symbol': symb,
+            'lastTradeDateOrContractMonth': date,
+            'strike': strike,
+            'right': right
+        }
+
+    def _convert_option_tots(self, ticker):
```
Comment (AdoNunes) [inline DiscordAlertsTrader/brokerages/ibkr_api.py:469]: _convert_option_to_ibrk

File: DiscordAlertsTrader/brokerages/ibkr_api.py [hunk 8]
```diff
@@ -0,0 +1,525 @@
...
+                    }]             
+
+                }
+                formatted_order = self.format_order(order_info)
+                return formatted_order
+    
+    def fix_symbol(self, symbol:str, direction:str):
+        "Fix symbol for options, direction in or out of webull format"
+        if direction == 'in':
+            return symbol.replace("SPXW", "SPX").replace("NDXP", "NDX")
```
Comment (AdoNunes) [inline DiscordAlertsTrader/brokerages/ibkr_api.py:212]: IBKR SPX is just SPX, not SPXW?

File: DiscordAlertsTrader/brokerages/ibkr_api.py (hunk 3 above)
Comment (Ish2K) [reply DiscordAlertsTrader/brokerages/ibkr_api.py:453]: Actually IBKR has the option as a Contract Object https://ib-insync.readthedocs.io/api.html#module-ib_insync.contract

I didn't realise it was meant to do that

File: DiscordAlertsTrader/brokerages/ibkr_api.py (hunk 4 above)
Comment (Ish2K) [reply DiscordAlertsTrader/brokerages/ibkr_api.py:327]: Need three parameters for entry, Take Profit and Sop Loss prices but only two are given, should add the entry price parameter? https://ib-insync.readthedocs.io/api.html#ib_insync.ib.IB.bracketOrder

Comment (Ish2K) [reply DiscordAlertsTrader/brokerages/ibkr_api.py:327]: Do you mean this? https://ib-insync.readthedocs.io/api.html#ib_insync.ib.IB.oneCancelsAll

File: DiscordAlertsTrader/brokerages/ibkr_api.py (hunk 6 above)
Comment (Ish2K) [reply DiscordAlertsTrader/brokerages/ibkr_api.py:368]: No actions can only take 'BUY' or 'SELL' as parameters. However, there is Limit on Close and Market on Close order types. Is that what you are looking for here?

File: DiscordAlertsTrader/brokerages/ibkr_api.py (hunk 8 above)
Comment (Ish2K) [reply DiscordAlertsTrader/brokerages/ibkr_api.py:212]: That's correct

File: DiscordAlertsTrader/brokerages/ibkr_api.py (hunk 6 above)
Comment (Ish2K) [reply DiscordAlertsTrader/brokerages/ibkr_api.py:368]: Reviewing code of other brokers, I don't think this is what you are looking for but please review and let me know 
https://interactivebrokers.github.io/tws-api/basic_orders.html#market_on_close
https://interactivebrokers.github.io/tws-api/basic_orders.html#limitonclose

Comment (AdoNunes) [reply DiscordAlertsTrader/brokerages/ibkr_api.py:387]: ok just making sure

**Pull request merged**